No fail when checksum file without Content-Length

setnicka opened the pull request:
When there is no Content-Length header the default value is -1. So when there is -1 we cannot do check of the response length.

cavaliercoder commented on client.go:
```diff
@@ -275,7 +275,7 @@ func (c *Client) checksumFile(resp *Response) stateFunc {
 		resp.err = err
 		return c.closeResponse
-	} else if nc != resp.Size {
+	} else if resp.Size != -1 && nc != resp.Size {
```
`resp.Size` is the sum of the `http.Response.ContentLength` and any bytes resumed from previous transfers. In the case that we are resuming a previous download, this is still broken. Thanks for highlighting this!

cavaliercoder commented:
Would you consider writing tests the cover the following cases?

* standard transfer where Content-Length is unknown
* resume previous transfer where Content-Length is unknown

setnicka commented:
I changed checks - now it directly checks HTTPResponse.ContentLength. In addition I implemented tests to check behavior for normal and resumed downloads.

I think it now should cover all cases (if I am not missing something).

setnicka commented:
Hello, would you find some time to review this pull request? :-)

cavaliercoder commented:
Thanks for your patience with this. Your code looks great, but there are so many possible states for this package - I need to be really cautious. I'm certain my tests are failing to cover some nuanced edge-case with this scenario. I'll try take a thorough look this weekend and get it merged.

setnicka commented:
Hi @cavaliercoder, something new about it? :-)

cavaliercoder commented:
Since rebasing, a couple of the checksum tests are failing, so we'll need to address that. Your test looks good, but I'd like to split it out from the existing `TestChecksums`. That test is more about how failed/valid checksums are handled, rather than ensuring a file can be resumed (tested elsewhere) or deal with edge-case server responses (this issue).

I'll adjust this and merge shortly! Thanks for the great work.

KennethNH commented:
When merge?

cavaliercoder commented on client_test.go:
```diff
@@ -70,55 +70,80 @@ func TestFilenameResolution(t *testing.T) {
+
+			// Normal request
+			req, _ = NewRequest(filename, ts.URL+fmt.Sprintf("?size=%d&sendlength=%v", test.size, test.sendlength))
 			req.SetChecksum(test.hash, b, true)
```
Reuse of `test-hash` here creates a data race between the two `DefaultClient.Do` calls.

cavaliercoder commented:
Hey @setnicka, I went ahead and fixed the underlying issue here in 8835fd25e1df4ff402859a6ea8fd1002d9f1d7f7.

I apologize for not merging your request here. I tried quite a few ways to resolve conflicts with some more recent refactors and fix some outstanding edge cases. In the end, it made more sense just to fix the issues by hand where they now live. As a small consolation, I've attributed the fix to you.
https://github.com/cavaliercoder/grab/commit/8835fd25e1df4ff402859a6ea8fd1002d9f1d7f7#diff-51565ad50d7426213c2f33579f6f0e71R712

Thanks for your understanding and excellent efforts here.
